core/src/apps/cardano/helpers/cbor_hash_builder.py

```diff
@@ -0,0 +1,87 @@
+from apps.cardano.helpers.lazy_cbor_collection import (
+    LazyCborCollection,
+    LazyCborDict,
+    LazyCborList,
+)
+
+if False:
+    from typing import Any, Protocol
+
+    class HashFn(Protocol):
+        def update(self, chunk: bytes) -> None:
+            ...
+
+        def digest(self) -> bytes:
+            ...
+
+
+class CborHashBuilder:
+    """
+    Builds a transaction hash in a streamed, lazy manner by keeping a stack of LazyCollections which get progressively
+    filled up with items.
+    """
+
+    def __init__(self, hash_fn: HashFn, initial_active_item: LazyCborCollection):
+        self.hash_fn = hash_fn
+        self.active_items_stack: list[LazyCborCollection] = [initial_active_item]
+        self.serializer = initial_active_item.cbor_serialize()
+
+        self._update_hash()
+
+    def _update_hash(self) -> None:
+        while (chunk := next(self.serializer)) is not LazyCborCollection.PauseIteration:
+            # for type checking
+            assert isinstance(chunk, bytes)
+            self.hash_fn.update(chunk)
+
+    def get_hash(self) -> bytes:
+        assert (
```

Comment:
Since the class is "protected" by `tx_builder` the asserts are IMHO safe and can only break by programmer error. However if we remove/update `tx_builder` as you propose then these might need to be actual errors.

Reply:
I replaced the asserts to 'if's raising a `RuntimeError` - I couldn't find a more suitable error. I've also added unit tests for the class. https://github.com/trezor/trezor-firmware/pull/1683/commits/fef973a76a34acf4d0118f2f078cb3382f78c6b6